pkg/tbtc/node.go

```diff
@@ -54,11 +56,14 @@ type node struct {
 	// signingExecutors is the cache holding signing executors for specific wallets.
 	// The cache key is the uncompressed public key (with 04 prefix) of the wallet.
 	signingExecutors map[string]*signingExecutor
```

Comment:
Having `walletDispatcher` we must ensure `signingExecutors` are not accessed directly outside of actions.

We already have one problem - `OnHeartbeatRequested` bypasses the `walletDispatcher` and calls a `signingExecutor` directly. As a result, we may have heartbeat and sweeping happening at the same time. This will only worsen when we realize multiple heartbeats can happen close to each other, given that a single machine can host multiple wallets.

Let's move the logic that is in `OnHeartbeatRequested` to a heartbeat wallet action. If the wallet is busy sweeping at the time the heartbeat should be executed, we should skip the heartbeat. Sweeping works as a heartbeat.

Reply:
I was thinking about it but there is a problem with heartbeats being a wallet action. Some time ago we agreed that an action is something complex that causes the change of wallet main UTXO while executors represent a lower-level action like signing. Hence, we have the `WalletAction` enum in the `WalletCoordinator` contract and its client-side representation `tbtc.WalletActionType`. Each `walletAction` has the `actionType` method that ties it to a specific `tbtc.WalletActionType` value. Now, heartbeats are just bare signing of random data and do not fulfill the definition of a wallet action. They are not managed by the `WalletCoordinator` and have no representation in the `tbtc.WalletActionType`. Having that in mind, things are more complex than they seem. 

I'm aware of the conflicts you mentioned but I'm not sure if making heartbeats a proper wallet action (despite of the aforementioned arguments against doing it) would help. If a heartbeat action is dispatched before a sweep it will block the sweep anyway and vice versa. I'm wondering whether we shouldn't use a kind of priority system. Something like "a sweep is more important than heartbeat" and "a redemption is more important than a sweep". And allow breaking the currently executed signing in favor of a higher-priority one. Not saying this is a good solution, just exploring possible ideas. WDYT?

Reply:
I think an action shouldn't necessarily change the UTXO but it should be _something the wallet is doing_. Heartbeat falls into this category. I think heartbeat could be a `WalletAction` in the `WalletCoordinator`.

It is not ideal that we'd coordinate heartbeats from the chain but it's also not ideal for sweeps, moving funds, and redemptions. Moreover, currently, only the most recent wallet executes heartbeats. We have to change it but it's also hard to expect all wallets to execute heartbeat at the same time. At least until we do not change the signing algorithm into something way less computationally-complex. How we'd enforce some order of heartbeats off-chain? That's the same problem as with coordinating sweeps - we'd have to reinvent the consensus mechanism. 

I would consider adding `Heartbeat` value to `WalletCoordinator`'s `WalletAction` and then having a separate `actionType` for it. We do not have to block this PR on it, but we should do it right after and leave a huge TODO block next to `chain.OnHeartbeatRequested(func(event *HeartbeatRequestedEvent)` in `tbtc.go`.

Does it make sense?

Reply:
Sure, we can change the action's definition in this way and pull the heartbeats under it. My concern was only about the fact that we need to treat all actions equally which means adding heartbeats to the `WalletCoordinator` as well. If we agree regarding this part that's good! If so, an additional question arises: how do we want to trigger the heartbeat? Just like sweeps using a function of `WalletCoordinator`?

Reply:
> If so, an additional question arises: how do we want to trigger the heartbeat? Just like sweeps using a function of WalletCoordinator?

I think yes, there is no point in complicating it and making special paths. 

The more we talk about it and the more problems we find I am keener to have the coordination on some L2 and a bot calling it. The cost will be lower than the dev time spent on the broadcast channel consensus.

Reply:
Seec 08769b5f

Reply:
Happened in https://github.com/keep-network/tbtc-v2/pull/607 and https://github.com/keep-network/tbtc-v2/pull/609.